src/lib/optimizer/strategy/expression_reduction_rule.cpp

```diff
@@ -243,4 +248,114 @@ void ExpressionReductionRule::rewrite_like_prefix_wildcard(std::shared_ptr<Abstr
+  for (auto& input_expression : input_expressions) {
+    if (input_expression->type != ExpressionType::Aggregate) continue;
+    auto& aggregate_expression = static_cast<AggregateExpression&>(*input_expression);
+    switch (aggregate_expression.aggregate_function) {
+      case AggregateFunction::Sum: {
+        sums.emplace_back(input_expression);
+        break;
+      }
+      case AggregateFunction::Count: {
+        counts.emplace_back(input_expression);
+        break;
+      }
+      case AggregateFunction::Avg: {
+        avgs.emplace_back(input_expression);
+        break;
+      }
+      default:
+        continue;
+    }
+  }
+
+  // Take a copy of what the aggregate was originally supposed to do
+  const auto original_aggregate_expressions = aggregate_node->column_expressions();
+
+  const auto& aggregate_input_node = aggregate_node->left_input();
+  auto replacements = ExpressionUnorderedMap<std::shared_ptr<AbstractExpression>>{};
+
+  // Iterate over the AVGs, check if matching SUMs and COUNTs exist, and add a suitable replacement to `replacements`.
+  for (const auto& avg_expression_ptr : avgs) {
+    const auto& avg_expression = static_cast<AggregateExpression&>(*avg_expression_ptr.get());
+
+    const auto& avg_argument = avg_expression.argument();
+    const auto avg_argument_is_nullable = avg_argument.get()->is_nullable_on_lqp(*aggregate_input_node);
+
+    // A helper function that checks whether SUMs and COUNTs match the AVG
+    const auto finder = [&](const auto& other_expression) {
+      const auto other_argument = static_cast<const AggregateExpression*>(other_expression.get().get())->argument();
+
+      if (!other_argument) {
+        // other_argument might be nullptr if we are looking at COUNT(*) - that is acceptable if the argument a to AVG
+        // is not nullable. In that case, COUNT(*) == COUNT(a).
+        return !avg_argument_is_nullable;
+      }
+
+      return other_argument == avg_argument || (other_argument && *other_argument == *avg_argument);
+    };
+
+    auto sum_it = std::find_if(sums.begin(), sums.end(), finder);
+    auto count_it = std::find_if(counts.begin(), counts.end(), finder);
+    if (sum_it != sums.end() && count_it != counts.end()) {
+      // Found matching SUM and COUNT (either COUNT(a) or COUNT(*) for a non-NULL a) - add it to the replacements list
+      // The cast will become unnecessary once #1799 is fixed
+      replacements[avg_expression_ptr] = div_(cast_(sum_it->get(), DataType::Double), count_it->get());
+    }
+  }
+
+  // No replacements possible
+  if (replacements.empty()) return;
+
+  // Back up the current column names
+  const auto& root_expressions = root_node->column_expressions();
+  auto old_column_names = std::vector<std::string>(root_expressions.size());
+  for (auto expression_idx = size_t{0}; expression_idx < root_expressions.size(); ++expression_idx) {
+    old_column_names[expression_idx] = root_expressions[expression_idx]->as_column_name();
+  }
+
+  {
```

Comment:
Why is that a separate scope?

Reply:
Because I did not want the name "expressions" to have a bigger scope. When it is possible without having to define variables before, I like small scopes.

Reply:
Understood.